File: Sources/AnagramEngine/PatternMatcher.swift
import Foundation

/// Crossword-style pattern search over a word list.
///
/// Pattern syntax:
///   - a letter a–z matches itself
///   - `?` or `.` matches exactly one letter
///   - `*` matches any run of letters (including none)
/// Matching is anchored to the whole word.
public struct PatternMatcher: Sendable {
    let wordList: WordList
    public init(wordList: WordList) { self.wordList = wordList }

    public func matches(pattern rawPattern: String, limit: Int = 0) -> [String] {
        let pattern = Array(rawPattern.lowercased())
        guard !pattern.isEmpty else { return [] }

        var out: [String] = []
        for w in wordList.words {
            if Self.match(Array(w), pattern) {
                out.append(w)
                if limit > 0 && out.count >= limit { break }
            }
        }
        return out
    }

    /// Glob-style match with `?`/`.` = one char and `*` = any run.
    static func match(_ word: [Character], _ pattern: [Character]) -> Bool {
        // Classic two-pointer wildcard matcher with backtracking on `*`.
        var wi = 0, pi = 0
        var star = -1, mark = 0
        while wi < word.count {
            if pi < pattern.count && (pattern[pi] == word[wi] || pattern[pi] == "?" || pattern[pi] == ".") {
                wi += 1; pi += 1
            } else if pi < pattern.count && pattern[pi] == "*" {
                star = pi; mark = wi; pi += 1
            } else if star != -1 {
                pi = star + 1; mark += 1; wi = mark
            } else {
                return false
            }
        }
        while pi < pattern.count && pattern[pi] == "*" { pi += 1 }
        return pi == pattern.count
    }
}

File: app/src/engine/patternMatcher.ts
// Ported from Sources/AnagramEngine/PatternMatcher.swift
import { WordList } from "./wordList";

/**
 * Crossword-style pattern search over a word list.
 *
 * Pattern syntax:
 *   - a letter a–z matches itself
 *   - `?` or `.` matches exactly one letter
 *   - `*` matches any run of letters (including none)
 * Matching is anchored to the whole word.
 */
export class PatternMatcher {
  private readonly wordList: WordList;

  constructor(wordList: WordList) {
    this.wordList = wordList;
  }

  matches(rawPattern: string, limit = 0): string[] {
    const pattern = rawPattern.toLowerCase();
    if (pattern.length === 0) return [];

    const out: string[] = [];
    for (const w of this.wordList.words) {
      if (PatternMatcher.match(w, pattern)) {
        out.push(w);
        if (limit > 0 && out.length >= limit) break;
      }
    }
    return out;
  }

  /** Glob-style match with `?`/`.` = one char and `*` = any run. */
  static match(word: string, pattern: string): boolean {
    // Classic two-pointer wildcard matcher with backtracking on `*`.
    let wi = 0,
      pi = 0;
    let star = -1,
      mark = 0;
    while (wi < word.length) {
      const p = pattern[pi];
      if (pi < pattern.length && (p === word[wi] || p === "?" || p === ".")) {
        wi++;
        pi++;
      } else if (pi < pattern.length && p === "*") {
        star = pi;
        mark = wi;
        pi++;
      } else if (star !== -1) {
        pi = star + 1;
        mark++;
        wi = mark;
      } else {
        return false;
      }
    }
    while (pi < pattern.length && pattern[pi] === "*") pi++;
    return pi === pattern.length;
  }
}
